chore: introduces MapStruct and uses it instead of manual mapping bet…

jekutzsche opened the pull request:
…ween DTOs and entities

The implementation has been done so far for the packages `users` and `vaccination_info`.

Also defines a central configuration with mapping for aggregate ids.

jekutzsche committed:
chore: introduces MapStruct and uses it instead of manual mapping between DTOs and entities

The implementation has been done so far for the packages `users` and `vaccination_info`.

Also defines a central configuration with mapping for aggregate ids.

lucky-lusa commented on iris-client-bff/src/main/java/iris/client_bff/vaccination_info/eps/VaccinationInfoControllerImpl.java:
```diff
@@ -98,45 +100,13 @@ private AnnouncementResultDto encryptAndCreateResult(VaccinationInfoAnnouncement
 
-	private VaccinationInfo.Facility map(Facility facilityDto) {
+	@Mapper(config = MapStructCentralConfig.class)
+	static interface DtoMapper {
```
Not a very speaking name - would put it into a dedicated file.

jekutzsche replied:
> Not a very speaking name - would put it into a dedicated file.

From my point of view, the mapper is only relevant within the controller and is an auxiliary class of it, just as the DTOs are a part of the interface. I therefore keep it with the Java documentation on nested classes:
> Why Use Nested Classes?
> 
> Compelling reasons for using nested classes include the following:
> 
> - It is a way of logically grouping classes that are only used in one place: If a class is useful to only one other class, then it is logical to embed it in that class and keep the two together. Nesting such "helper classes" makes their package more streamlined.
> - It increases encapsulation: Consider two top-level classes, A and B, where B needs access to members of A that would otherwise be declared private. By hiding class B within class A, A's members can be declared private and B can access them. In addition, B itself can be hidden from the outside world.
> - It can lead to more readable and maintainable code: Nesting small classes within top-level classes places the code closer to where it is used.
> (https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html)

Logically grouping, encapsulation, a better overview and maintenance of useful visibility rules are for me strong arguments for nested classes.

In the context of the `VaccinationInfoControllerImpl` I find the name then quite suitable, since `VaccinationInfoControllerImpl` as a "controller" takes over a rather clear task. At this point, however, we could also make two mappers `FacilityMapper` and `EmployeeMapper`.

lucky-lusa reviewed (commented):
We should take a look at naming and mappers placement.

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/users/web/UserProfileController.java:
```diff
@@ -18,14 +16,13 @@
-			return UserMappers.map(byUsername.get(), userService);
-		}
+
+		var userOpt = userService.findByUsername(principal.getName());
```
That's right. The suffix `Opt` only makes sense if I need value in optional also explicitly as variable.

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/users/web/UserMapper.java:
```diff
@@ -0,0 +1,17 @@
+
+	@Mapping(target = "createdBy", qualifiedBy = UserNameMetadata.class)
+	@Mapping(target = "lastModifiedBy", qualifiedBy = UserNameMetadata.class)
+	UserDTO mapEntity2Dto(UserAccount source);
```
I think it is better to name the source of the mapping as well. With `toDto` and `toEntity` you have to assume that the source is an entity and a DTO respectively, but that doesn't always have to be the case. So I wanted to use clearer names that leave fewer questions.

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/vaccination_info/eps/VaccinationInfoControllerImpl.java:
```diff
@@ -31,7 +31,8 @@
 	private final EncryptionService encryptionService;
-	private final ObjectMapper mapper;
+	private final ObjectMapper objectMapper;
+	private final DtoMapper mapper;
```
The last point is understandable. I like to shorten the variable names and fields within a certain range, since the type is also available as info (class names are often not the shortest). But since there are two mappers here, it is probably better longer.

jekutzsche commented on iris-client-bff/src/main/java/iris/client_bff/vaccination_info/web/VaccinationInfoController.java:
```diff
@@ -38,6 +40,7 @@
 public class VaccinationInfoController {
 
 	private final VaccinationInfoService service;
+	private final DtoMapper mapper;
```
Yes, same as other controller. 🙂

jekutzsche committed:
chore: improves naming

jekutzsche committed:
chore: improves mapper testing by avoiding error-prone mocking

Introduces an annotation (with the necessary helper classes) for mapper tests that can be used to implement a Spring test that makes error-prone mocking unnecessary with the Spring context. Thereby the Spring context is built specifically for the mappers with general base mappers and the Spring Conversion Service including all converters to avoid long execution times with the whole context.

jekutzsche commented:
:tada: This PR is included in version 1.6.0-rc.1 :tada:

The release is available on [GitHub release](https://github.com/iris-connect/iris-client/releases/tag/v1.6.0-rc.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

jekutzsche commented:
:tada: This PR is included in version 1.6.0 :tada:

The release is available on [GitHub release](https://github.com/iris-connect/iris-client/releases/tag/v1.6.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
